lake/commits/patch.go

```diff
@@ -174,16 +201,70 @@ func (p *Patch) Revert(tip *Snapshot, commit, parent ksuid.KSUID, retries int, a
 	return object, nil
 }
 
-func (p *Patch) OverlappingDeletes(with *Patch) []ksuid.KSUID {
-	lookup := make(map[ksuid.KSUID]struct{})
-	for _, id := range p.deletedObjects {
-		lookup[id] = struct{}{}
+func Diff(parent, child *Patch) (*Patch, error) {
+	var dirty bool
+	p := NewPatch(parent)
+	deletedObjects := make(map[ksuid.KSUID]struct{})
+	for _, id := range child.deletedObjects {
+		deletedObjects[id] = struct{}{}
 	}
-	var ids []ksuid.KSUID
-	for _, id := range with.deletedObjects {
-		if _, ok := lookup[id]; ok {
-			ids = append(ids, id)
+	// For each object in the child patch that isn't in the parent, create an add,
+	// unless the parent deletes it, then return an error.
+	for _, o := range child.SelectAll() {
+		if !Exists(parent, o.ID) {
+			if _, ok := deletedObjects[o.ID]; ok {
+				return nil, fmt.Errorf("parent branch deletes object that child branch adds: %d", o.ID)
+			}
+			if err := p.AddDataObject(o); err != nil {
+				return nil, err
+			}
+			dirty = true
 		}
 	}
-	return ids
+	// For each delete in the child patch, create a delete.
+	// If the object doesn't exist in the parent, then we have a
+	// delete conflict.
+	for _, id := range child.deletedObjects {
+		if Exists(parent, id) {
+			if err := p.DeleteObject(id); err != nil {
+				return nil, err
+			}
+			dirty = true
+		} else {
+			return nil, fmt.Errorf("delete conflict: %s", id)
+		}
+	}
+	deletedIndexes := make(map[string]struct{})
+	for _, idx := range child.deletedIndexes {
+		deletedIndexes[idx.Key()] = struct{}{}
+	}
+	// For each index entry in child that isn't in the parent, create an add-index.
+	for _, idx := range child.SelectAllIndexes() {
+		if !IndexExists(parent, idx.Rule.RuleID(), idx.ID) {
+			key := indexRef{id: idx.ID, ruleID: idx.Rule.RuleID()}.Key()
+			if _, ok := deletedIndexes[key]; ok {
+				return nil, fmt.Errorf("parent branch deletes index object that child branch adds: %s:%s", idx.Rule.RuleID(), idx.ID)
+			}
+			if err := p.AddIndexObject(idx); err != nil {
+				return nil, err
+			}
+			dirty = true
+		} else {
+			return nil, fmt.Errorf("delete conflict on index: %s:%s", idx.Rule.RuleID(), idx.ID)
+		}
+	}
+	// For each delete-index in the child patch, create a delete-index.
+	// (XXX should return an error if it's not in the parent).
+	for _, idx := range child.deletedIndexes {
+		if IndexExists(parent, idx.ruleID, idx.id) {
+			if err := p.DeleteIndexObject(idx.ruleID, idx.id); err != nil {
+				return nil, err
+			}
+			dirty = true
+		}
+	}
+	if !dirty {
+		return nil, errors.New("nothing to merge")
```

Comment:
"nothing to merge" seems like the wrong error message given the function is called Diff.

Reply:
Maybe address this by changing the function name to Merge?

Reply:
I'll do "difference is empty" then this error is decorated by the merge logic with additional context so that works nicely.